bout-particle-push.cxx

```diff
@@ -0,0 +1,767 @@
+  // output << "Got here -1 \n";
+
+  PETSCCHK(PetscInitializeNoArguments());
+  auto sycl_target = std::make_shared<SYCLTarget>(0, PETSC_COMM_WORLD);
+  const int mpi_size = sycl_target->comm_pair.size_parent;
+  const int mpi_rank = sycl_target->comm_pair.rank_parent;
+  // output << "Got here 0 \n";
+  // global number of physical nonunique vertices stored in hypnotoad datasets
+  const int N_nonunique_vertices = mpi_size*Nx*Ny;
+  // arrays to fill with local data
+  std::vector<double> local_Z_lower_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_R_lower_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_Z_lower_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_R_lower_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_Z_upper_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_R_upper_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_Z_upper_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> local_R_upper_left_vertices(N_nonunique_vertices, 0.0);
+  // arrays to receive the summed data across ranks
+  std::vector<double> global_Z_lower_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_R_lower_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_Z_lower_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_R_lower_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_Z_upper_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_R_upper_right_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_Z_upper_left_vertices(N_nonunique_vertices, 0.0);
+  std::vector<double> global_R_upper_left_vertices(N_nonunique_vertices, 0.0);
+  // fill these vectors with vertex values from the local rank
+  // at indices determined by the local rank
+  int icxy = Nx*Ny*mpi_rank;
+  for (int ix = mesh->xstart; ix<= mesh->xend; ix++) {
+    for (int iy = mesh->ystart; iy <= mesh->yend; iy++) {
+      local_R_lower_left_vertices.at(icxy) = Rxy_lower_left_corners(ix,iy);
+      local_Z_lower_left_vertices.at(icxy) = Zxy_lower_left_corners(ix,iy);
+      local_R_lower_right_vertices.at(icxy) = Rxy_lower_right_corners(ix,iy);
+      local_Z_lower_right_vertices.at(icxy) = Zxy_lower_right_corners(ix,iy);
+      local_R_upper_right_vertices.at(icxy) = Rxy_upper_right_corners(ix,iy);
+      local_Z_upper_right_vertices.at(icxy) = Zxy_upper_right_corners(ix,iy);
+      local_R_upper_left_vertices.at(icxy) = Rxy_upper_left_corners(ix,iy);
+      local_Z_upper_left_vertices.at(icxy) = Zxy_upper_left_corners(ix,iy);
+      icxy++;
+    }
+  }
+  // Perform Allreduce (sum) to get knowledge of vertices to all ranks
+  MPICHK(MPI_Allreduce(local_R_lower_left_vertices.data(), global_R_lower_left_vertices.data(), N_nonunique_vertices, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD));
```

Comment:
Remember to run clang-format using the Hermes-3 .clang-format definition regularly otherwise you end up with large git diffs when it is eventually run.

Reply:
addressed by https://github.com/UKAEA-Edge-Code/hermes-3/pull/1/commits/8e678d2aaa0beec49b5ab6c690a983876f7ab936